extensions/2.0/Khronos/KHR_materials_iridescence/README.md

```diff
@@ -0,0 +1,207 @@
+This effect is the result of constructive and destructive interferences of certain wavelengths. 
+If the the optical path difference between the ray reflected at the thin-film and the ray reflected at the base material is half the wavelength (λ), the resulting 180 degree phase shift is cancelling out the reflected light:
+
+<figure>
+<img width=70% src="./figures/interference.png"/>
+<figcaption><em> </em></figcaption>
+</figure>
+
+With a thin-film thickness near half the wavelength of visible light (380 nm - 750 nm), the effect is most visible. 
+By increasing the thin-film thickness, multiples of wavelengths are still causing wave interferences, however, as the optical path distance increases, different rays are mixed in. 
+This leads to more pastel colored patterns for increased thickness:
+
+<figure>
+<img src="./figures/thickness-comparison.jpg"/>
+<figcaption><em>Comparison of different iridescence thickness ranges for a constant iridescence IOR value of 1.8. </em></figcaption>
+</figure>
+
+The thin-film layer can have a different IOR than the underlying material. With `iridescenceIOR` one can set an IOR value for the thin-film layer independently. 
+This also has an effect on the optical path difference:
+
+<figure>
+<img src="./figures/ior-comparison.jpg"/>
+<figcaption><em>Comparison of different iridescence IOR values for a thin-film thickness range between 200 nm (bottom) and 800 nm (top).</em></figcaption>
+</figure>
+
+
+The iridescence effect is modeled via a microfacet BRDF with a modified Fresnel reflectance term that accounts for inter-reflections as shown in [Laurent Belcour and Pascal Barla (2017)](https://belcour.github.io/blog/research/publication/2017/05/01/brdf-thin-film.html).
+
+
+## Implementation Notes
+
+*This section is non-normative.*
+
+For the calculation of `f_specular`, the original Fresnel term `F_schlick` is exchanged with a newly introduced `F_iridescence` term. 
+To artistically control the strength of the iridescence effect, the two Fresnel terms are mixed by using the `iridescenceFactor`:
+```
+F = mix(F_schlick, F_iridescence, iridescenceFactor)
+```
+
+To calculate `F_iridescence` we need to calculate the spectral differences. 
+At first, optical path differences (`diff`) between rays reflected at the thin-film and at the base material have to be calculated:
+```
+diff = 2.0 * IOR_iridescence * thin_film_thickness * cos(viewAngle)
+```
+With the path difference, the phase shift can be calculated in Fourier space.
+
+<img src="https://render.githubusercontent.com/render/math?math=\displaystyle%20\Delta%20\phi%20=%202%20\pi%20\lambda^{-1}\text{diff}">
+
+A complete implementation for evaluating `F_iridescence` can be found in the shader file of the Khronos Sample Viewer:
+[iridescence.glsl](https://github.com/PascalSchoen/glTF-Sample-Viewer/blob/feature/KHR_materials_iridescence/source/Renderer/shaders/iridescence.glsl)
```

Comment:
Some notes on the `fresnelConductorExact` function which is being used in the sample code:

```
vec3 eta_1 = vec3(eta1);

// Force eta_2 -> 1.0 when Dinc -> 0.0
vec3 eta_2 = mix(eta_1, vec3(eta2), smoothstep(0.0, 0.03, Dinc));

vec3 eta_3, kappa_3;
artisticIor(baseF0, mix(vec3(0.0), baseF0, metallic), eta_3, kappa_3);

// ...

fresnelConductorExact(cosTheta1, eta_2[i] / eta_1[i], 0.0, R12p[i], R12s[i]);
fresnelConductorExact(cosTheta2[i], eta_3[i] / eta_2[i], kappa_3[i] / eta_2[i], R23p[i], R23s[i]);
```

First call is first interface, second call is second interface (if I am not mistaken). The name implies that this is a Fresnel for conductors (I found a *very* similar looking function in [Mitsuba](https://github.com/mitsuba-renderer/mitsuba/blob/master/src/libcore/util.cpp#L739), in case you took it from there, note that this is GPL license). However, for the first interface, I would expect a dielectric Fresnel. You even mention this in the property table: "iridescenceIOR - The index of refraction of the dielectric thin-film layer." For the second interface, I would expect it to be either dielectric or conductor, depending on the metalness (`metallic` parameter) of the base material.

`eta_3` and `kappa_3`, i.e., the complex IOR of the base, is computed in `artisticIor`. There are several issues:
* `artisticIor` implements Gulbrandsen's Fresnel term. I am wondering why you chose that term. It's not used in the glTF spec, and its use is actually discouraged because it has several problems. See (Naty Hoffman's talk "Fresnel Equations Considered Harmful")[https://renderwonk.com/publications/mam2019/] or the discussions in KHR_materials_specular for the reasons.
* The `mix` in the call to `artisticIor` sets `edgeColor` to `baseF0` if `metallic = 1`. This is not how the glTF base material behaves. According to the base spec, F90 = 1 for metals.

In summary, I think the Fresnel at the first interface should be a regular dielectric Fresnel (maybe Schlick, maybe the full Fresnel equations) and the Fresnel at the second interface should be `F` from the base spec, KHR_materials_ior and KHR_materials_specular (i.e., it correctly accounts for metallic, baseColor, specular, specular color, and volume IOR).